document-portal/document-portal-fuse.c

```diff
@@ -2460,23 +2521,48 @@ xdp_fuse_unlink (fuse_req_t  req,
   xdp_reply_err (op, req, 0);
 }
 
+static int
+try_renameat (int olddirfd,
+              const char *oldpath,
+              int newdirfd,
+              const char *newpath,
+              unsigned int flags)
+{
+#if HAVE_RENAMEAT2
+  return renameat2 (olddirfd, oldpath, newdirfd, newpath, flags);
+#else
+  if (flags)
+    {
+      g_critical ("renameat2 is not supported by this system and rename flags are set");
+      errno = EINVAL;
```

Comment:
If the system genuinely doesn't have `renameat2()` then `ENOSYS` would be more realistic.

Reply:
The kernel clearly has a `renameat2` system call if the FUSE kernel module is sending us renameat2 calls with non-zero flags. The issue is that without a renameat2 function in glibc, we can't easily implement the missing behaviour (without e.g. coding our own syscall wrappers or trying to reimplement the same semantics).

Also, renameat2() calls with flags==0 or for paths on different file systems will likely succeed.

Reply:
I'm not sure what's your suggestion here :smile:, as it's consistent but still prefer to just warn?

Reply:
I think it might be worth reducing the warning severity: it's a useful debugging information, but doesn't indicate a catastrophic failure. From the application side, it's a failure mode that should be expected and handled with alternative code (and the EINVAL error seems like the right code to indicate that the file system doesn't support it).